File: 973.k-closest-points-to-origin.java
import java.sql.Time;
import java.util.Comparator;
import java.util.PriorityQueue;

/*
 * @lc app=leetcode id=973 lang=java
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start

// class Solution {
//     public int[][] kClosest(int[][] points, int k) {
//         PriorityQueue<int[]> pq = new PriorityQueue<int[]>(
//                 (p1, p2) -> p2[0] * p2[0] + p2[1] * p2[1] - p1[0] * p1[0] - p1[1] * p1[1]);

//         for (int i = 0; i < points.length; i++) {
//             pq.add(points[i]);
//             if (pq.size() > k) {
//                 pq.poll();
//             }
//         }

//         int[][] res = new int[k][];
//         for (int i = 0; i < k; i++) {
//             res[i] = pq.poll();
//         }

//         return res;
//     }
// }

// Average case: pivot is generally near the mid of l and r; 
// therefore we have n + (n/2 + n/4 +...) <= n + (n) = 2 * n = O(n).
// As for worst case: K = n, and points is already sorted by the distance to origin. 
// @Frimish's code will cost n - 1 + ... + 1 = (n - 1) * n / 2 = O(n^2) time.
class Solution {
    // since return the answer in any order
    // we can consider quick sort (quick select)
    public int[][] kClosest(int[][] points, int K) {
        int len = points.length, l = 0, r = len - 1;
        while (l <= r) {
            // one round quick select
            int mid = helper(points, l, r);
            if (mid == K)
                break;
            if (mid < K) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return Arrays.copyOfRange(points, 0, K);
    }

    private int helper(int[][] A, int l, int r) {
        // pick A[l] as pivot,
        // let every e <= pivot on the left side of pivot and e > pivot on the right
        int[] pivot = A[l];
        while (l < r) {
            while (l < r && compare(A[r], pivot) >= 0) {
                r--;
            }
            A[l] = A[r];
            while (l < r && compare(A[l], pivot) <= 0) {
                l++;
            }
            A[r] = A[l];
        }
        A[l] = pivot;
        return l;
    }

    private int compare(int[] p1, int[] p2) {
        return p1[0] * p1[0] + p1[1] * p1[1] - p2[0] * p2[0] - p2[1] * p2[1];
    }
}

// @lc code=end

File: 973.k-closest-points-to-origin.py
#
# @lc app=leetcode id=973 lang=python3
#
# [973] K Closest Points to Origin
#

# @lc code=start

# class Solution:
#     def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
#         # defaultly a max heap
#         heap = []

#         for i in range(len(points)):
#             dis = -(points[i][0]**2+points[i][1]**2)
#             if len(heap) == k:
#                 # guaranteed to be unique
#                 if dis > heap[0][0]:
#                     heapq.heapreplace(heap, (dis, i))
#             else:
#                 heapq.heappush(heap, (dis, i))

#         return [points[i] for (_, i) in heap]

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        # quick_select, since we can return in any order
        """Perform the QuickSelect algorithm on the list"""
        left, right = 0, len(points) - 1
        pivot_index = len(points)
        while pivot_index != k:
            # Repeatedly partition the list
            # while narrowing in on the kth element
            pivot_index = self.partition(points, left, right)
            if pivot_index < k:
                left = pivot_index
            else:
                right = pivot_index - 1

        # Return the first k elements of the partially sorted list
        return points[:k]

    def partition(self, points: List[List[int]], left: int, right: int) -> int:
        """Partition the list around the pivot value"""
        """Choose a pivot element of the list"""
        pivot = points[left + (right - left) // 2]
        pivot_dist = self.squared_distance(pivot)
        while left < right:
            # Iterate through the range and swap elements to make sure
            # that all points closer than the pivot are to the left
            if self.squared_distance(points[left]) >= pivot_dist:
                points[left], points[right] = points[right], points[left]
                right -= 1
            else:
                left += 1

        # Ensure the left pointer is just past the end of
        # the left range then return it as the new pivotIndex
        if self.squared_distance(points[left]) < pivot_dist:
            left += 1
        return left

    def squared_distance(self, point: List[int]) -> int:
        """Calculate and return the squared Euclidean distance."""
        return point[0] ** 2 + point[1] ** 2

# @lc code=end
